### PR TITLE
Okio filesystem (3.0 alpha)

### PR DESCRIPTION
**Pull request by yschimke**
Releasing based on alpha Okio dependency

### CODE REVIEW

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt [hunk 1]
```diff
@@ -132,27 +141,35 @@ class Main : Callable<Int> {
...
 
-  internal fun addHostToCompletionFile(host: String) {
+  @Suppress("BlockingMethodInNonBlockingContext")
+  internal suspend fun addHostToCompletionFile(host: String) {
     val previousHosts = knownHosts()
     val newHosts = previousHosts + host
 
     val lineSeparator = System.getProperty("line.separator")
-    knownHostsFile.writeText(newHosts.joinToString(lineSeparator, postfix = lineSeparator))
+    filesystem.sink(knownHostsFile.also { filesystem.createDirectories(it.parent!!) }).buffer().use {
```
Comment (swankjesse) [inline certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt:158]: the lack of a createDirectories call on the old code is surprising!

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt [hunk 2]
```diff
@@ -165,16 +182,15 @@ class Main : Callable<Int> {
   companion object {
     internal const val NAME = "cft"
 
-    val confDir = File(System.getProperty("user.home"), ".cft").also {
-      if (!it.isDirectory) {
```
Comment (swankjesse) [inline certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt:169]: why remove this?

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/certs.kt [hunk 3]
```diff
@@ -35,18 +40,29 @@ internal fun String.parsePemCertificate(fileName: String? = null): Certificate {
...
-  output: File
+@Suppress("BlockingMethodInNonBlockingContext")
+@OptIn(ExperimentalFilesystem::class)
+internal suspend fun X509Certificate.writePem(
+  output: Path,
+  filesystem: Filesystem = Filesystem.SYSTEM
 ) {
-  output.writeText(certificatePem())
+  withContext(Dispatchers.IO) {
+    filesystem.sink(output).buffer().use {
```
Comment (swankjesse) [inline certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/certs.kt:64]: I think we probably want `filesystem.writeUtf8(output, certificatePem())`

**Review by swankjesse — APPROVED** [review]
Neat! Definitely shows the lack of sugar in our APIs

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt (hunk 2 above)
Comment (yschimke) [reply certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt:169]: Trying to use Path API exclusively, and at this point we don't know the Filesystem.

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt (hunk 1 above)
Comment (yschimke) [reply certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/Main.kt:158]: Any suggestions?

File: certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/certs.kt (hunk 3 above)
Comment (yschimke) [reply certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/certs.kt:64]: extension methods would be great here.

Comment (yschimke) [reply certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/certs.kt:64]: on the flip side.

```
fun Path.writeUtf8(fs: Filesystem = Filesystem.SYSTEM) {}
```

Would be problematic but nice for the defaults.

**Comment by yschimke** [comment]
> Neat! Definitely shows the lack of sugar in our APIs

Should I riff on some sugary APIs here?  Any suggestions? including suspend functions? high level (safe) operations that don't require closing?

**Pull request merged**